File: mergeSort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void intercala(int inicio, int meio, int fim, int v[], int tam){
    int inicio_v1 = inicio;
    int inicio_v2 = meio+1;
    int pos_livre = 0;
    int aux[tam];

    while(inicio_v1<=meio && inicio_v2<=fim){
        if(v[inicio_v1] <= v[inicio_v2])
            aux[pos_livre++] = v[inicio_v1++];
        else
            aux[pos_livre++] = v[inicio_v2++];
    }

    while(inicio_v1 <= meio)
        aux[pos_livre++] = v[inicio_v1++];

    while(inicio_v2 <= fim)
        aux[pos_livre++] = v[inicio_v2++];

    for(inicio_v1 = inicio; inicio_v1 <= fim; inicio_v1++)
        v[inicio_v1] = aux[inicio_v1-inicio];
}

void mergeSort(int inicio, int fim, int v[], int tam){
    if(inicio<fim){
        int meio = (inicio+fim)/2;
        mergeSort(inicio, meio,v, tam);
        mergeSort(meio+1, fim, v, tam);
        intercala(inicio, meio, fim, v, tam);
    }
}

void executar(int tam, int caso, FILE *f){
    double tempos_execucao[1000];

    for(int j=0; j<1000; j++){
        int v[tam];
        clock_t inicio, fim;
        double tempo;

        for(int i=0; i<tam; i++){
            if (caso == 1){
                v[i] = i;
            }
            else if (caso == 2){
                v[i] = rand();
            }
            else if (caso == 3){
                v[i] = tam - i;
            }
        }

        inicio = clock();
        mergeSort(0, tam-1, v, tam);
        fim = clock();

        tempo = ((double)(fim - inicio)) / CLOCKS_PER_SEC;
        tempos_execucao[j] = tempo;
    }

    const char *nome_caso;
    if (caso == 1) nome_caso = "Melhor";
    else if (caso == 2) nome_caso = "Caso medio";
    else nome_caso = "Pior";

    for(int i=0; i<1000; i++){
        fprintf(f, "%s,%d,%d,%.9f\n", nome_caso, tam, i+1, tempos_execucao[i]);
    }
}

int main(void){
    FILE *f = fopen("tempos_c.csv", "w");
    if(f == NULL){
        printf("Erro ao abrir arquivo para escrita\n");
        return 1;
    }

    fprintf(f, "Caso,Tamanho,Execucao,Tempo\n");

    executar(100, 1, f);
    executar(100, 2, f);
    executar(100, 3, f);

    executar(1000, 1, f);
    executar(1000, 2, f);
    executar(1000, 3, f);

    executar(10000, 1, f);
    executar(10000, 2, f);
    executar(10000, 3, f);

    fclose(f);
    return 0;
}

File: mergeSort.py
import time
import random

def mergeSort(inicio: int, fim: int, v: list[int], tam: int):
    if(inicio < fim):
        meio = (inicio+fim)//2
        mergeSort(inicio, meio, v, tam)
        mergeSort(meio+1, fim, v, tam)
        intercala(inicio, meio, fim, v, tam)

def intercala(inicio: int, meio:int, fim: int, v:list[int], tam: int):
    inicio_v1 = inicio
    inicio_v2 = meio+1
    pos_livre = 0
    aux = [0] * tam

    while(inicio_v1 <= meio and inicio_v2 <= fim):
        if(v[inicio_v1] <= v[inicio_v2]):
            aux[pos_livre] = v[inicio_v1]
            pos_livre+=1
            inicio_v1+=1
        else:
            aux[pos_livre] = v[inicio_v2]
            inicio_v2+=1
            pos_livre+=1
    
    while(inicio_v1 <= meio):
        aux[pos_livre] = v[inicio_v1]
        inicio_v1+=1
        pos_livre+=1
    
    while(inicio_v2 <= fim):
        aux[pos_livre] = v[inicio_v2]
        inicio_v2 +=1
        pos_livre+=1
    
    for i in range(pos_livre):
        v[inicio + i] = aux[i]


# para os vetores que tem tamanho 100
tempos_melhor_100 = []
tempos_pior_100 = []
tempos_medio_100 = []

for i in range(1000):
    v1 = list(range(100))                    # melhor caso
    v2 = list(range(100, 0, -1))             # pior caso
    v3 = [random.randint(1,100000) for _ in range(100)]  # caso médio

    for v, lista_tempos in [(v1, tempos_melhor_100), (v2, tempos_pior_100), (v3, tempos_medio_100)]:
        tam = len(v)
        tempo_inicial = time.time()
        mergeSort(0,tam-1,v,tam)
        tempo_final = time.time()
        lista_tempos.append(tempo_final - tempo_inicial)


# para os vetores que têm tamanho 1000
tempos_melhor_1000 = []
tempos_pior_1000 = []
tempos_medio_1000 = []
for i in range(1000):
    v1 = list(range(1000))                    # melhor caso
    v2 = list(range(1000, 0, -1))             # pior caso
    v3 = [random.randint(1,100000) for _ in range(1000)]  # caso médio

    for v, lista_tempos in [(v1, tempos_melhor_1000), (v2, tempos_pior_1000), (v3, tempos_medio_1000)]:
        tam = len(v)
        tempo_inicial = time.time()
        mergeSort(0,tam-1,v,tam)
        tempo_final = time.time()
        lista_tempos.append(tempo_final-tempo_inicial)

# para os vetores que têm tamanho 10000
tempos_melhor_10000 = []
tempos_pior_10000 = []
tempos_medio_10000 = []
for i in range(1000):
    v1 = list(range(10000))                    # melhor caso
    v2 = list(range(10000, 0, -1))             # pior caso
    v3 = [random.randint(1,100000) for _ in range(10000)]  # caso médio

    for v, lista_tempos in [(v1, tempos_melhor_10000), (v2, tempos_pior_10000), (v3, tempos_medio_10000)]:
        tam = len(v)
        tempo_inicial = time.time()
        mergeSort(0,tam-1,v,tam)
        tempo_final = time.time()
        lista_tempos.append(tempo_final - tempo_inicial)


print("\nTAMANHO 100")
print("Média melhor caso:", sum(tempos_melhor_100)/10)
print("Média pior caso:  ", sum(tempos_pior_100)/10)
print("Média caso médio: ", sum(tempos_medio_100)/10)

print("\nTAMANHO 1000")
print("Média melhor caso:", sum(tempos_melhor_1000)/10)
print("Média pior caso:  ", sum(tempos_pior_1000)/10)
print("Média caso médio: ", sum(tempos_medio_1000)/10)

print("\nTAMANHO 10000")
print("Média melhor caso:", sum(tempos_melhor_10000)/10)
print("Média pior caso:  ", sum(tempos_pior_10000)/10)
print("Média caso médio: ", sum(tempos_medio_10000)/10)
